port/nordic/nrf5x/src/hal/spim.zig

```diff
@@ -0,0 +1,323 @@
+                .M16 => .M16,
+                .M32 => .M32,
+            } }),
+        }
+
+        regs.ORC.write(.{ .ORC = config.overread_char });
+
+        config.sck_pin.set_direction(.out);
+        config.sck_pin.set_drive_strength(config.sck_drive);
+        switch (compatibility.chip) {
+            .nrf52 => regs.PSEL.SCK.write(.{
+                .PIN = config.sck_pin.index(),
+                .CONNECT = .Connected,
+            }),
+            .nrf52840 => regs.PSEL.SCK.write(.{
+                .PIN = config.sck_pin.index(),
+                .PORT = config.sck_pin.port(),
+                .CONNECT = .Connected,
+            }),
+        }
+
+        regs.ENABLE.write(.{ .ENABLE = .Enabled });
+        regs.INTENCLR.write_raw(0xFFFFFFFF);
+        // TODO: Enable interrupts
+    }
+
+    pub fn write_blocking(spi: SPIM, data: []const u8, timeout: ?mdf.time.Duration) TransactionError!void {
+        const deadline = mdf.time.Deadline.init_relative(time.get_time_since_boot(), timeout);
+        try spi.inner_tx_rx(data, &.{}, deadline);
+    }
+
+    pub fn writev_blocking(spi: SPIM, chunks: []const []const u8, timeout: ?mdf.time.Duration) TransactionError!void {
+        if (timeout) |initial_timeout| {
+            var remaining_timeout = initial_timeout;
+            var start_time = time.get_time_since_boot();
+
+            for (chunks) |chunk| {
+                const elapsed_time = time.get_time_since_boot().diff(start_time);
+                remaining_timeout = remaining_timeout.minus(elapsed_time);
+
+                try spi.write_blocking(chunk, remaining_timeout);
+                start_time = time.get_time_since_boot();
+            }
+        } else {
+            for (chunks) |chunk| {
+                try spi.write_blocking(chunk, null);
+            }
+        }
+    }
+
+    pub fn read_blocking(spi: SPIM, dst: []u8, timeout: ?mdf.time.Duration) TransactionError!void {
+        const deadline = mdf.time.Deadline.init_relative(time.get_time_since_boot(), timeout);
+        try spi.inner_tx_rx(&.{}, dst, deadline);
+    }
+
+    pub fn readv_blocking(spi: SPIM, chunks: []const []u8, timeout: ?mdf.time.Duration) TransactionError!void {
+        if (timeout) |initial_timeout| {
+            var remaining_timeout = initial_timeout;
+            var start_time = time.get_time_since_boot();
+
+            for (chunks) |chunk| {
+                const elapsed_time = time.get_time_since_boot().diff(start_time);
+                remaining_timeout = remaining_timeout.minus(elapsed_time);
+
+                try spi.read_blocking(chunk, remaining_timeout);
+                start_time = time.get_time_since_boot();
+            }
+        } else {
+            for (chunks) |chunk| {
+                try spi.read_blocking(chunk, null);
+            }
+        }
+    }
+
+    pub fn transceivev_blocking(spi: SPIM, write_chunks: []const []const u8, read_chunks: []const []u8, timeout: ?mdf.time.Duration) TransactionError!void {
```

Comment:
This function basically doesn't work for DMA. I think that maybe I just don't write it here, and MAYBE the datagram device can choose to use the non-dma HAL (once written) for this use case.

Another options is that I make a buf with the total rx len, loop, and then copy in the chunks after?